calyx/backend/src/verilog.rs

```diff
@@ -722,6 +730,242 @@ fn cell_instance(cell: &ir::Cell) -> Option<v::Instance> {
+            let guard_unmet_value = if is_data_port(&destination) {
+                "'x".to_string()
+            } else {
+                format!("{}'d0", destination.borrow().width)
+            };
+
+            writeln!(
+                f,
+                " {} ? {} :",
+                case_guarded_assign_guard,
+                VerilogPortRef(&assignment.src)
+            )?;
+
+            if i + 1 == num_merged_assigns {
+                writeln!(f, " {guard_unmet_value};")?;
+            }
+        }
+    }
+
+    io::Result::Ok(())
+}
+
+fn emit_fsm_module<F: io::Write>(
+    fsm: &RRC<ir::FSM>,
+    comp_name: ir::Id,
+    f: &mut F,
+) -> io::Result<()> {
+    let num_states = fsm.borrow().assignments.len();
+    let reg_bitwidth = get_bit_width_from(num_states as u64);
+
+    // Write module header. Inputs include ports checked during transitions, and
+    // outputs include one one-bit wire for every state
+    writeln!(f, "\nmodule {}_{comp_name}_def (", fsm.borrow().name())?;
+    writeln!(f, "  input logic clk,")?;
+    writeln!(f, "  input logic reset,")?;
+    let mut used_port_names: HashSet<ir::Canonical> = HashSet::new();
+    for transition in fsm.borrow().transitions.iter() {
+        if let ir::Transition::Conditional(guards) = transition {
+            for (guard, _) in guards.iter() {
+                for port in guard.all_ports().iter() {
+                    if used_port_names.insert(port.borrow().canonical()) {
+                        let wire_width = match port.borrow().width {
+                            1 => "".to_string(),
+                            n => format!("[{}:{}]", n - 1, 0),
+                        };
+                        writeln!(
+                            f,
+                            "  input logic {} {},",
+                            wire_width,
+                            VerilogPortRef(port)
+                        )?;
+                    }
+                }
+            }
+        }
+    }
+    for state in 0..num_states {
+        writeln!(
+            f,
+            "  output logic s{}_out{}",
+            state,
+            if state < num_states - 1 { "," } else { "" }
+        )?;
+    }
+    writeln!(f, ");\n")?;
+
+    // Write symbolic state variables and give them binary implementations
+    for state in 0..num_states {
+        writeln!(f, "  parameter s{state} = {reg_bitwidth}'d{state};")?;
+    }
+
+    writeln!(f)?;
+
+    // State register logic variable
+    writeln!(f, "  logic [{}:0] state_reg;", reg_bitwidth - 1)?;
+    writeln!(f, "  logic [{}:0] state_next;\n", reg_bitwidth - 1)?;
+
+    // Generate sequential block representing the FSM
+    writeln!(f, "  always @(posedge clk) begin")?;
+    writeln!(f, "    if (reset) begin")?;
+    writeln!(f, "      state_reg <= s0;")?;
+    writeln!(f, "    end")?;
+    writeln!(f, "    else begin")?;
+    writeln!(f, "      state_reg <= state_next;")?;
+    writeln!(f, "    end")?;
+    writeln!(f, "  end\n")?;
```

Comment:
Can you make this one long raw string (using `"""`) instead of multiple `writeln!` calls?